toolkit/tools/precacher/precacher.go

```diff
@@ -0,0 +1,409 @@
+	availablePackages, err := getAllRepoData(*repoUrls, *workerTar, *buildDir)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	logger.Log.Infof("Pre-caching: Found %d available packages", len(availablePackages))
+	if logger.Log.IsLevelEnabled(logrus.DebugLevel) {
+		for _, pkg := range availablePackages {
+			logger.Log.Debugf("Found package: %s", pkg)
+		}
+	}
+
+	// For each package in the snapshot, check if it is available online and try to download it
+	downloadedPackages, err := downloadMissingPackages(rpmSnapshot, availablePackages, *outDir, *concurrentNetOps)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	logger.Log.Infof("Pre-caching: Downloaded %d packages into the cache", len(downloadedPackages))
+	err = writeSummaryFile(*outputSummaryFile, downloadedPackages)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+}
+
+func rpmSnapshotFromFile(snapshotFile string) (rpmSnapshot *repocloner.RepoContents, err error) {
+	err = jsonutils.ReadJSONFile(snapshotFile, &rpmSnapshot)
+	return
+}
+
+// getAllRepoData returns a map of package names to URLs for all packages available in the given repos. It uses
+// a chroot to run repoquery.
+func getAllRepoData(repoUrls []string, workerTar, buildDir string) (namesToUrls map[string]string, err error) {
+	const (
+		leaveChrootOnDisk = false
+	)
+	timestamp.StartEvent("pull available package data from repos", nil)
+	defer timestamp.StopEvent(nil)
+
+	// Create a chroot to run repoquery in
+	// Create the directory 'buildDir' if it does not exist
+	exists, err := file.DirExists(buildDir)
+	if err != nil {
+		err = fmt.Errorf("failed to check if directory %s exists: %w", buildDir, err)
+		return nil, err
+	}
+	if !exists {
+		logger.Log.Infof("Pre-caching: Creating 1st time chroot directory %s", buildDir)
+		err = os.MkdirAll(buildDir, 0755)
+		if err != nil {
+			err = fmt.Errorf("failed to create directory %s: %w", buildDir, err)
+			return nil, err
+		}
+	}
+	queryChroot, err := createChroot(workerTar, buildDir, leaveChrootOnDisk)
+	if err != nil {
+		err = fmt.Errorf("failed to create chroot: %w", err)
+		return nil, err
+	}
+	defer queryChroot.Close(leaveChrootOnDisk)
+
+	namesToUrls = make(map[string]string)
+	for _, repoUrl := range repoUrls {
+		// Use the chroot to query each repo for the packages it contains
+		var packages []string
+		err = queryChroot.Run(func() (chrootErr error) {
+			packages, chrootErr = getRepoPackages(repoUrl)
+			return chrootErr
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		// We will be searching by the name: "<name>-<version>.<distro>.<arch>", the results from the repoquery will be
+		// in the form of "<PARTIAL_URL>/<name>-<version>.<distro>.<arch>.rpm"
+		for _, pkgUrl := range packages {
+			name := path.Base(pkgUrl)
+			name = strings.TrimSuffix(name, ".rpm")
+
+			// We need to prepend the repoUrl to the partial URL to get the full URL
+			pkgUrl = fmt.Sprintf("%s/%s", repoUrl, pkgUrl)
+			namesToUrls[name] = pkgUrl
+		}
+	}
```

Comment:
OK, now I see we don't retrieve full URLs but just the path component of the URL with `getRepoPackage`. In that case, how about:
```suggestion
	for _, repoURL := range repoURLs {
		// Use the chroot to query each repo for the packages it contains
		var packages []string
		err = queryChroot.Run(func() (chrootErr error) {
			packageRepoPaths, chrootErr = getPackageRepoPaths(repoURL)
			return chrootErr
		})
		if err != nil {
			return nil, err
		}

		// We will be searching by the name: "<name>-<version>.<distro>.<arch>", the results from the repoquery will be
		// in the form of "<PARTIAL_URL>/<name>-<version>.<distro>.<arch>.rpm"
		for _, packageRepoPath := range packageRepoPaths {
			packageName := path.Base(packageRepoPath)
			packageName = strings.TrimSuffix(packageName, ".rpm")

			// We need to prepend the repoURL to the partial URL to get the full URL
			pkgURL = fmt.Sprintf("%s/%s", repoURL, packageRepoPath)
			namesToUrls[name] = pkgURL
		}
	}
```